### PR TITLE
DOC-12574: 3.2.1 Maintenance Release

### PR DESCRIPTION
**Pull request by ElliotFrancisHunter**
PR for Sync Gateway 3.2.1 maintenance release. Outside of typical docs requirements there is also an addition to the what's new page and the db config schema. Linked below.

https://preview.docs-test.couchbase.com/maint321/home/index.html

Preview Credentials: https://couchbasecloud.atlassian.net/wiki/spaces/DOCS/pages/1971224949/Docs+Team+demo+site+passwords

Changed Pages
-------------------

- What's new: https://preview.docs-test.couchbase.com/maint321/sync-gateway/current/whatsnew.html
- DB config change: https://preview.docs-test.couchbase.com/maint321/sync-gateway/current/configuration-schema-database.html#cache-rev_cache-max_memory_count_mb
- Revisions doc - https://preview.docs-test.couchbase.com/maint321/sync-gateway/current/revisions.html#lbl-max-memory
- Release notes: https://preview.docs-test.couchbase.com/maint321/sync-gateway/current/release-notes.html#maint-latest

----------------------
Pages still needing changes/updates
-----------------------

....

-----------------------

### CODE REVIEW

**Review by osfameron — APPROVED** [review]
Looks good to me

File: modules/ROOT/pages/whatsnew.adoc [hunk 1]
```diff
@@ -25,6 +25,13 @@ This can only happen during a mixed mode of SGW with a 3.2 version and an older
 
 --
 
+* *Additional Rev Cache Configuration Options*
+
+You can now specify a memory limit on your rev cache with the `rev_cache.max_memory_count_mb` configuration option.
+You can use this option alongside existing configurable limits to the number of documents in the rev cache to reduce the risk of OOM issues.
+
+For more information, see xref:configuration-schema-database.adoc#cache-rev_cache-max_memory_count_mb[]
```
Comment (idulo) [inline modules/ROOT/pages/whatsnew.adoc]: @ElliotFrancisHunter this link does not present with a user friendly text, e.g. see rev cache memory limit config, but with an internal structure with the #. Can that be changed?

**Review by idulo — APPROVED** [review]
Approval upon resolving one comment

File: modules/ROOT/pages/whatsnew.adoc (hunk 1 above)
Comment (ElliotFrancisHunter) [reply modules/ROOT/pages/whatsnew.adoc]: It seems antora cannot automatically populate xrefs with user friendly text from the anchored header. Easy fix is to manually add the text in the square brackets myself, good spot!

Comment (ElliotFrancisHunter) [reply modules/ROOT/pages/whatsnew.adoc]: Change is in! Updating preview now

**Comment by idulo** [comment]
@ElliotFrancisHunter We do not have the corresponding changes in the Revisions and the [Revisions Cache](https://docs.couchbase.com/sync-gateway/current/revisions.html#lbl-caching) sections. This is a P0 to explain new rev cache configuration, concept and impact to users in the relevant docs section.

File: modules/ROOT/assets/attachments/bundled-admin.yaml [hunk 2]
```diff
@@ -3928,6 +3928,13 @@ paths:
                             description: The maximum number of revisions that can be stored in the revision cache.
                             type: string
                             default: 5000
+                          max_memory_count_mb:
+                            description: |-
+                              The maximum amount of memory the revision cache should take up in MB, setting to 0 will disable any eviction based on memory at rev cache.
+                              When set this memory limit will work in in hand with revision cache size parameter. So you will potentially get eviction at revision cache both based off memory footprint and number of items in the cache.
+                              **This is an enterprise-edition feature only**
```
Comment (idulo) [inline modules/ROOT/assets/attachments/bundled-admin.yaml]: @ElliotFrancisHunter We always seem to refer to EE as Enterprise Edition with capitals, see [here](https://docs.couchbase.com/server/current/introduction/editions.html).

File: modules/ROOT/pages/revisions.adoc [hunk 3]
```diff
@@ -175,18 +175,20 @@ Whenever a document is accessed its revision tree (or at least some portion of i
...
+* xref:configuration-schema-database.adoc#cache-rev_cache-max_memory_count_mb[rev_cache.max_memory_count_mb]
+
 * {configuration-schema-database--xref--cache--rev-cache--shard-count}
 
+* {configuration-schema-database--xref--cache--rev-cache--size}
 
-[#lbl-size]
-=== Size
-Use the {configuration-schema-database--xref--cache--rev-cache--size} setting to specify the total number of document revisions to be cached in memory for all (recently accessed) documents.
+[#lbl-max-memory]
```
Comment (idulo) [inline modules/ROOT/pages/revisions.adoc]: @ElliotFrancisHunter this config should be colocated with the other rev cache size config, size. Shouldn't have sharding which is unrelated in between them.

These two configs work hand in hand as it is very well explained in the db config page: 

> When set this memory limit will work in in hand with revision cache size parameter. So you will potentially get eviction at revision cache both based off memory footprint and number of items in the cache. 

We should communicate to users what the connection between those two configs is (controlling size of rev cache) and the implications for setting both that is outlined in the quote above. Maybe even have them colocated in a subsection. Thoughts?

**Review by idulo — CHANGES_REQUESTED** [review]

File: modules/ROOT/assets/attachments/bundled-admin.yaml (hunk 2 above)
Comment (ElliotFrancisHunter) [reply modules/ROOT/assets/attachments/bundled-admin.yaml]: Done!

File: modules/ROOT/pages/_partials/release_notes/sync-gateway-3-2-1-release-note.adoc [hunk 4]
```diff
@@ -0,0 +1,44 @@
+== 3.2.1 -- October 2024
+
+[#maint-3-2-1]
+=== Fixed Issues
+
+* https://jira.issues.couchbase.com/browse/CBG-4075[CBG-4075 -  Timeout error will release document sequcne as unused]
```
Comment (bbrks) [inline modules/ROOT/pages/_partials/release_notes/sync-gateway-3-2-1-release-note.adoc]:
```suggestion
* https://jira.issues.couchbase.com/browse/CBG-4075[CBG-4075 -  Timeout error will release document sequence as unused]
```

File: modules/ROOT/pages/_partials/release_notes/sync-gateway-3-2-1-release-note.adoc [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+* https://jira.issues.couchbase.com/browse/CBG-4151[CBG-4151 - Memory-based rev cache size tuning]
+
+* https://jira.issues.couchbase.com/browse/CBG-4170[CBG-4170 - Stat for rev cache current capacity]
+
+* https://jira.issues.couchbase.com/browse/CBG-4174[CBG-4174 - Force disconnection of blip clients on database close]
+
+* https://jira.issues.couchbase.com/browse/CBG-4176[CBG-4176 -  Improved Auth Failed Logging]
+
+* https://jira.issues.couchbase.com/browse/CBG-4189[CBG-4189 - Add metric for Assertions]
```
Comment (bbrks) [inline modules/ROOT/pages/_partials/release_notes/sync-gateway-3-2-1-release-note.adoc]: This one got moved to 3.2.2 - see linked jira ticket

File: modules/ROOT/pages/revisions.adoc [hunk 6]
```diff
@@ -175,9 +175,22 @@ Whenever a document is accessed its revision tree (or at least some portion of i
...
+* {configuration-schema-database--xref--cache--rev-cache--size}
+
+[#lbl-max-memory]
+=== Maximum Memory
+
+include::partial$block-caveats.adoc[tags=ee-only-feature]
+
+You can specify the maximum amount of memory the revision cache should take up in MB, setting the value to `0` will disable any eviction based on memory at rev cache. 
+
+When set, this maximum memory limit will work with the `cache.rev_cache.size` parameter, allowing you to configure limits to the number of documents and the maximum memory usage within the rev cache to reduce the risk of OOM issues.
```
Comment (bbrks) [inline modules/ROOT/pages/revisions.adoc]: Slight wording tweak to clarify the interaction between both settings, but I am open to other opinions.

```suggestion
When set, this maximum memory limit will work in conjunction with the `cache.rev_cache.size` parameter, allowing you to configure limits to the number of documents and the maximum memory usage within the rev cache to reduce the risk of OOM issues.
```

**Review by bbrks — DISMISSED** [review]
Just a couple of things - give or take, but the mentioned 3.2.2 issue will need removing.

File: modules/ROOT/pages/revisions.adoc (hunk 3 above)
Comment (ElliotFrancisHunter) [reply modules/ROOT/pages/revisions.adoc]: Done

File: modules/ROOT/pages/_partials/release_notes/sync-gateway-3-2-1-release-note.adoc (hunk 5 above)
Comment (ElliotFrancisHunter) [reply modules/ROOT/pages/_partials/release_notes/sync-gateway-3-2-1-release-note.adoc]: Good spot! Removed.

File: modules/ROOT/pages/_partials/release_notes/sync-gateway-3-2-1-release-note.adoc (hunk 4 above)
Comment (ElliotFrancisHunter) [reply modules/ROOT/pages/_partials/release_notes/sync-gateway-3-2-1-release-note.adoc]: Fixed

File: modules/ROOT/pages/revisions.adoc (hunk 6 above)
Comment (ElliotFrancisHunter) [reply modules/ROOT/pages/revisions.adoc]: Happy to make that change. Done!

**Review by osfameron — APPROVED** [review]
👍

File: modules/ROOT/pages/revisions.adoc [hunk 7]
```diff
@@ -175,19 +175,31 @@ Whenever a document is accessed its revision tree (or at least some portion of i
...
+
 * {configuration-schema-database--xref--cache--rev-cache--shard-count}
 
+* {configuration-schema-database--xref--cache--rev-cache--size}
+
+[#lbl-cache-limit-config]
+=== Cache Limit Configuration
+
+There are two ways to configure the size of the revision cache: through memory in MB using xref:configuration-schema-database.adoc#cache-rev_cache-max_memory_count_mb[`rev_cache.max_memory_count_mb`], and through size in number of document revisions in cache xref:configuration-schema-database.adoc#cache-rev_cache_size[`rev_cache.size`]. 
+These configurations allow you to configure both limits to the number of documents and the maximum memory usage within the rev cache to reduce the risk of OOM issues.
```
Comment (osfameron) [inline modules/ROOT/pages/revisions.adoc]: "OOM issues" isn't defined or linked.

Comment (osfameron) [reply modules/ROOT/pages/revisions.adoc]: (here, later, and in whatsnew.adoc)

File: modules/ROOT/pages/revisions.adoc [hunk 8]
```diff
@@ -175,19 +175,31 @@ Whenever a document is accessed its revision tree (or at least some portion of i
...
+* xref:configuration-schema-database.adoc#cache-rev_cache-max_memory_count_mb[rev_cache.max_memory_count_mb]
+
 * {configuration-schema-database--xref--cache--rev-cache--shard-count}
 
+* {configuration-schema-database--xref--cache--rev-cache--size}
+
+[#lbl-cache-limit-config]
+=== Cache Limit Configuration
+
+There are two ways to configure the size of the revision cache: through memory in MB using xref:configuration-schema-database.adoc#cache-rev_cache-max_memory_count_mb[`rev_cache.max_memory_count_mb`], and through size in number of document revisions in cache xref:configuration-schema-database.adoc#cache-rev_cache_size[`rev_cache.size`].
```
Comment (simon-dew) [inline modules/ROOT/pages/revisions.adoc]: Tightening up the wording

```suggestion
There are two ways to configure the size of the revision cache: set a memory limit in MB using xref:configuration-schema-database.adoc#cache-rev_cache-max_memory_count_mb[`rev_cache.max_memory_count_mb`], or set the maximum number of document revisions in cache using xref:configuration-schema-database.adoc#cache-rev_cache_size[`rev_cache.size`]. 
```

File: modules/ROOT/pages/revisions.adoc [hunk 9]
```diff
@@ -175,19 +175,31 @@ Whenever a document is accessed its revision tree (or at least some portion of i
...
-[#lbl-size]
-=== Size
 Use the {configuration-schema-database--xref--cache--rev-cache--size} setting to specify the total number of document revisions to be cached in memory for all (recently accessed) documents.
 
 When the revision cache is full, Sync Gateway will remove older document revisions to make room for newer ones.
 
 By adjusting this setting you can fine-tune Sync Gateway's memory consumption.
 This can be useful when working on servers with limited memory and in cases when Sync Gateway creates and-or updated many new documents relative to the number of read operations.
 
+You can use the xref:configuration-schema-database.adoc#cache-rev_cache-max_memory_count_mb[`rev_cache.max_memory_count_mb`] setting to specify the maximum amount of memory the revision cache should take up in MB, setting the value to `0` will disable any eviction based on memory at rev cache.
```
Comment (simon-dew) [inline modules/ROOT/pages/revisions.adoc]: Punctuation typo

```suggestion
You can use the xref:configuration-schema-database.adoc#cache-rev_cache-max_memory_count_mb[`rev_cache.max_memory_count_mb`] setting to specify the maximum amount of memory the revision cache should take up in MB.
Setting the value to `0` will disable any eviction based on memory at rev cache. 
```

**Review by simon-dew — APPROVED** [review]
Generally looks okay (quick and dirty review) but have some suggestions to tighten up wording and punctuation.

**Pull request merged**